### PR TITLE
8283699: Improve the peephole mechanism of hotspot

### PR DESCRIPTION
**Pull request by merykitty**
Hi,

The current peephole mechanism has several drawbacks:
- Can only match and remove adjacent instructions.
- Cannot match machine ideal nodes (e.g MachSpillCopyNode).
- Can only replace 1 instruction, the position of insertion is limited to the position at which the matched nodes reside.
- Is actually broken since the nodes are not connected properly and OptoScheduling requires true dependencies between nodes.

The patch proposes to enhance the peephole mechanism by allowing a peep rule to call into a dedicated function, which takes the responsibility to perform all required transformations on the basic block. This allows the peephole mechanism to perform several transformations effectively in a more fine-grain manner.

The patch uses the peephole optimisation to perform some classic peepholes, transforming on x86 the sequences:

    mov r1, r2    ->    lea r1, [r2 + r3/i]
    add r1, r3/i

and

    mov r1, r2    ->    lea r1, [r2 << i], with i = 1, 2, 3
    shl r1, i

On the added benchmarks, the transformations show positive results:

    Benchmark             Mode  Cnt     Score     Error  Units
    LeaPeephole.B_D_int   avgt    5  1200.490 ± 104.662  ns/op
    LeaPeephole.B_D_long  avgt    5  1211.439 ±  30.196  ns/op
    LeaPeephole.B_I_int   avgt    5  1118.831 ±   7.995  ns/op
    LeaPeephole.B_I_long  avgt    5  1112.389 ±  15.838  ns/op
    LeaPeephole.I_S_int   avgt    5  1262.528 ±   7.293  ns/op
    LeaPeephole.I_S_long  avgt    5  1223.820 ±  17.777  ns/op

    Benchmark             Mode  Cnt    Score    Error  Units
    LeaPeephole.B_D_int   avgt    5  860.889 ±  6.089  ns/op
    LeaPeephole.B_D_long  avgt    5  945.455 ± 21.603  ns/op
    LeaPeephole.B_I_int   avgt    5  849.109 ±  9.809  ns/op
    LeaPeephole.B_I_long  avgt    5  851.283 ± 16.921  ns/op
    LeaPeephole.I_S_int   avgt    5  976.594 ± 23.004  ns/op
    LeaPeephole.I_S_long  avgt    5  936.984 ±  9.601  ns/op

A following patch would add IR tests for these transformations since the IR framework has not been able to parse the ideal scheduling yet although printing the scheduling itself has been made possible recently.

Thank you very much.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8283699](https://bugs.openjdk.org/browse/JDK-8283699): Improve the peephole mechanism of hotspot

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)
 * [Dean Long](https://openjdk.org/census#dlong) (@dean-long - **Reviewer**) ⚠️ Review applies to [566b8dd1](https://git.openjdk.org/jdk/pull/8025/files/566b8dd13d9f9ff382630eef8463cf0bc1199afc)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/8025/head:pull/8025` \
`$ git checkout pull/8025`

Update a local copy of the PR: \
`$ git checkout pull/8025` \
`$ git pull https://git.openjdk.org/jdk pull/8025/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8025`

View PR using the GUI difftool: \
`$ git pr show -t 8025`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/8025.diff">https://git.openjdk.org/jdk/pull/8025.diff</a>

</details>

### CODE REVIEW

**Comment by merykitty** [comment]
In case this has not reached the mailing list, may someone take a look at this, please.
Thank you very much.

**Review by vnkozlov — COMMENTED** [review]
Very interesting. Let me test it.

**Comment by vnkozlov** [comment]
There were failure in testing. I reported it in JBS.

**Comment by vnkozlov** [comment]
Unfortunately the only information I have is in JBS comment. I would only look on TestBase64.java failure. It was run on Oracle Linux and Ice Lake Intel's CPU. It could be intermittent failure which is not related to your changes but I don't see any previous records similar to this.

**Comment by vnkozlov** [comment]
I ran previously built binaries and TestBase64.java failure happened 10 out of 10 runs. It is not intermittent.

I attached hs_err file to RFE. I created it with `-XX:AbortVMOnException=java.lang.IndexOutOfBoundsException`

**Comment by merykitty** [comment]
Thanks a lot for your testing, can you run the tests again, please?

**Comment by vnkozlov** [comment]
@merykitty I started testing for version 03.

**Review by vnkozlov — APPROVED** [review]
Testing passed.

**Comment by vnkozlov** [comment]
You need second review.

**Comment by merykitty** [comment]
That's great, thanks very much for your review.

File: src/hotspot/cpu/x86/peephole_x86.cpp
Comment (dean-long) [inline src/hotspot/cpu/x86/peephole_x86.cpp:28]: I don't see why opto/peephole.hpp is useful.  Why not just include peephole_x86.hpp?  Then the empty peephole_<cpu>.hpp for the other platforms are no longer needed.

File: src/hotspot/cpu/x86/peephole_x86.cpp
Comment (dean-long) [inline src/hotspot/cpu/x86/peephole_x86.cpp:50]: I don't understand why this optimization requires MachSpillCopy.  Is that the only time we sould see mov+add or mov+shift?

File: src/hotspot/cpu/x86/peephole_x86.cpp
Comment (merykitty) [reply src/hotspot/cpu/x86/peephole_x86.cpp:132]: While it seems to be a little boiler-plate, I think a general helper that removes, inserts an arbitrary number of nodes, and connects the graph correctly is hard to write. What do you think? Thanks a lot.

File: src/hotspot/cpu/x86/peephole_x86.cpp
Comment (dean-long) [reply src/hotspot/cpu/x86/peephole_x86.cpp:50]: OK.

File: src/hotspot/cpu/x86/peephole_x86.cpp
Comment (dean-long) [reply src/hotspot/cpu/x86/peephole_x86.cpp:132]: How does it work when there is no peepprocedure?  I would expect that all the boiler-plate details are taken care of by peepreplace.

File: src/hotspot/cpu/x86/peephole_x86.cpp
Comment (dean-long) [reply src/hotspot/cpu/x86/peephole_x86.cpp:28]: How about including `peephole_<cpu>.hpp` only when "peepprocedure" is seen, and delete opto/peephole.hpp and empty `peephole_<cpu>.hpp` files?

File: src/hotspot/cpu/x86/peephole_x86.cpp
Comment (merykitty) [reply src/hotspot/cpu/x86/peephole_x86.cpp:132]: The code produced by `peepreplace` is really limited, it cannot be applied in this case because both the `MachSpillCopy` and `MachProj` are not match rules. Also, I think a generated code is more powerful than a helper function, so achieving something general with a helper function would be hard. I think we can refactor it into a helper if there are similarities arise later instead. What do you think?

File: src/hotspot/cpu/x86/peephole_x86.cpp
Comment (merykitty) [reply src/hotspot/cpu/x86/peephole_x86.cpp:28]: What do you think if I include `peephole_x86.hpp` in `x86_64.ad` in a source hpp block? This will result in the include appearing in `ad_x86.hpp`, which will be transitively included in `ad_x86_peephole.cpp`. Thanks a lot.

Comment (dean-long) [reply src/hotspot/cpu/x86/peephole_x86.cpp:28]: Yes, good idea.

File: src/hotspot/cpu/x86/peephole_x86.cpp
Comment (dean-long) [reply src/hotspot/cpu/x86/peephole_x86.cpp:132]: OK, agreed.

File: src/hotspot/cpu/x86/peephole_x86.cpp
Comment (iwanowww) [reply src/hotspot/cpu/x86/peephole_x86.cpp:28]: Considering `peephole_x86.cpp` contains  only x64-specific code, I had a suggestion to rename it into `peephole_x86_64.cpp` and move `#ifdef _LP64` into `peephole_x86.hpp` to guard x64-specific declarations. But if you intend to include the header directly from `x86_64.ad`, you can rename the header to `peephole_x86_64.hpp` and get rid of `#ifdef _LP64` completely.

Comment (merykitty) [reply src/hotspot/cpu/x86/peephole_x86.cpp]: I have refactored that and removed `opto/peephole.hpp` as well as other peephole headers. `peephole_x86.cpp` is also renamed to `peephole_x86_64.cpp`. Thanks very much.

**Review by dean-long — APPROVED** [review]

**Review by vnkozlov — COMMENTED** [review]
I started build testing. But I can only verify 64-bit.
@merykitty Can you verify 32 build too?

File: src/hotspot/cpu/x86/x86_64.ad
Comment (iwanowww) [inline src/hotspot/cpu/x86/x86_64.ad:324]: Why don't you simply include `peephole_x86_64.hpp` here?

Comment (merykitty) [reply src/hotspot/cpu/x86/x86_64.ad]: Done

**Comment by merykitty** [comment]
@vnkozlov I saw GHA being happy with x86_32, I also tried cross-compiling locally for 32-bit build.

**Comment by vnkozlov** [comment]
> @vnkozlov I saw GHA being happy with x86_32, I also tried cross-compiling locally for 32-bit build.

Good. My builds in tier1 also passed. I don't have any more comments.

**Comment by merykitty** [comment]
Thank you very much for your reviews!
/integrate

**Comment by merykitty** [comment]
@vnkozlov Could you reapprove this PR, please?

**Review by vnkozlov — APPROVED** [review]
Good.

**Comment by merykitty** [comment]
Thanks a lot
/integrate

**Pull request closed without merging**
